src/layer/rail.js

```diff
@@ -0,0 +1,497 @@
+"use strict";
+
+import * as Color from "../constants/color.js";
+import * as Util from "../js/util.js";
+
+// Exponent base for inter-zoom interpolation
+let railExp = 1.2;
+
+// Helper functions to create zoom interpolation expressions
+function multiplyMatchExpression(value, factor) {
+  if (Array.isArray(value)) {
+    var result = [value[0], value[1]];
+    for (let i = 2; i < value.length - 1; i++) {
+      if (i % 2 == 0) {
+        result.push(value[i]);
+      } else {
+        result.push(multiplyMatchExpression(value[i], factor));
+      }
+    }
+    result.push(multiplyMatchExpression(value[value.length - 1], factor));
+    return result;
+  } else {
+    return value * factor;
+  }
+}
+
+function zoomInterpolate(widthZ20) {
+  return [
+    "interpolate",
+    ["exponential", railExp],
+    ["zoom"],
+    8,
+    multiplyMatchExpression(widthZ20, 1 / 16),
+    12,
+    multiplyMatchExpression(widthZ20, 1 / 4),
+    20,
+    widthZ20,
+  ];
+}
+
+// Helper function to create a "filter" block for a particular railway class.
+function filterRail(brunnel) {
+  return [
+    "all",
+    brunnel === "surface"
+      ? ["!in", "brunnel", "bridge", "tunnel"]
+      : ["==", "brunnel", brunnel],
+    ["in", "class", "rail", "transit"],
+  ];
+}
+
+// Base definition that applies to all railways
+var defRail = {
+  type: "line",
+  source: "openmaptiles",
+  "source-layer": "transportation",
+};
+
+var serviceSelector = ["match", ["get", "service"], ["siding", "spur", "yard"]];
+var isService = ["in", "service", "siding", "spur", "yard"];
+var isNotService = ["!in", "service", "siding", "spur", "yard"];
+
+var lineColor = [
+  "match",
+  ["get", "brunnel"],
+  "tunnel",
+  "hsl(0, 0%, 90%)",
```

Comment:
I think it would be helpful to pull out these magic colors into the Color class.  As the style grows, it will help us to examine the color palette holistically.